# 1、竞赛背景
如何有效利用海量的广告数据和用户数据去预测用户的广告点击概率，是大数据应用在精准营销中的关键问题。   
本次大赛提供了讯飞AI营销云的海量广告投放数据，参赛选手通过人工智能技术构建预测模型预估用户的广告点击概率，即给定广告点击相关的广告、媒体、用户、上下文内容等信息的条件下预测广告点击概率。

# 2、评分标准
## 评分算法
other
## 评分标准
![](http://third.datacastle.cn/pkbigdata/master.other.img/6c1e634c-897a-4e84-bfac-2f5c4d440009.png)

## AB榜的划分方式和比例：
【1】评分采用AB榜形式。排行榜显示A榜成绩，竞赛结束后2小时切换成B榜单。B榜成绩以选定的两次提交或者默认的最后两次提交的最高分为准，最终比赛成绩以B榜单为准。   
【2】此题目的AB榜是随机划分，A榜为随机抽样测试集50%数据，B榜为另外50%。

# 3、任务
讯飞AI营销广告点击率预估，预测广告被点击的概率。

# 4、数据
提供下载的数据集包括两个部分：  
初赛：
1. round1_iflyad_train.txt 训练集，每一行数据为一个样本，可分为5类数据，包含基础广告投放数据、广告素材信息、媒体信息、用户信息和上下文信息，共1001650条数据。其中‘click’字段为要预测的标签，其它34个字段为特征字段。
2. round1_iflyad_test_feature.txt 测试集，共40024条数据，与训练集文件相比，测试集文件无‘click’字段，其它字段同训练集。

复赛：
1. round2_iflyad_train.txt 训练集，每一行数据为一个样本，可分为5类数据，包含基础广告投放数据、广告素材信息、媒体信息、用户信息和上下文信息，共1998350条数据。其中‘click’字段为要预测的标签，其它34个字段为特征字段。 
2. round2_iflyad_test_feature.txt 测试集，共80276条数据，与训练集文件相比，测试集文件无‘click’字段，其它字段同训练集。

出于数据安全保证的考虑，所有数据均为脱敏处理后的数据。数据集提供了若干天的样本，最后一天数据构成了测试集，其余日期的数据作为训练数据。

注意：此次比赛分为初赛和复赛两个阶段，两个阶段的区别是所提供样本的量级有所不同，其他的设置均相同。

基础广告投放数据（1+1）、广告素材信息（16）、媒体信息（5）、用户信息（1）、上下文信息（11）

# 1、导入依赖包

In [2]:
import numpy as np            # ndarray数组
import pandas as pd           # DataFrame表格
from sklearn import datasets  # 自带的数据集
from sklearn.model_selection import train_test_split # 随机划分为训练子集和测试子集
from sklearn.model_selection import cross_val_score  # 模型评价：训练误差和测试误差
from sklearn.feature_selection import SelectFromModel# 特征选择(三种方法)
from sklearn.metrics import roc_auc_score            # 评价指标
from sklearn.metrics import f1_score 
from sklearn.cross_validation import StratifiedKFold # K折交叉验证

from sklearn.neighbors import KNeighborsClassifier   # KNN
from sklearn.linear_model import LogisticRegression  # 逻辑斯特回归LR
from sklearn.tree import DecisionTreeClassifier      # DT
from sklearn.ensemble import RandomForestClassifier  # RFC随机森林分类
from sklearn.ensemble import RandomForestRegressor   # RFR随机森林回归
from sklearn.ensemble import ExtraTreesClassifier    # ETC极端随机树分类
from sklearn.ensemble import ExtraTreesRegressor     # ETR极端随机树回归
from sklearn.naive_bayes import GaussianNB           # GNB朴素贝叶斯
from sklearn import svm                              # SVM支持向量机
import xgboost as xgb                                # XGB
import lightgbm as lgb                               # LGB

import matplotlib as mpl
import matplotlib.pyplot as plt      # 作图
import seaborn as sns                # 作图

from IPython.display import display  # 输出语句
plt.style.use("fivethirtyeight")
sns.set_style({'font.sans-serif':['simhei','Arial']})

import warnings                      # 消除警告
warnings.filterwarnings("ignore")
import os             # 系统模块
%matplotlib inline  

# 深度学习框架
import mxnet
import tensorflow as tf

# 检查Python版本
from sys import version_info
if version_info.major != 3:
    raise Exception('请使用 Python 3 来完成此项目')

# 2、加载数据集

In [3]:
# 加载数据
train = pd.read_table('./data/round1_iflyad_train.txt')
test = pd.read_table('./data/round1_iflyad_test_feature.txt')

# 合并训练集，验证集
data = pd.concat([train,test], axis=0, ignore_index=True)

data.head()

adid  advert_id advert_industry_inner       advert_name  app_cate_id  \
0  1560128  230000063         102400_102401  B4734117F35EE97F        107.0   
1  1488859  230000063         102400_102401  B4734117F35EE97F        108.0   
2  1537089  230000065         101700_101704  E257895F74792E81        100.0   
3  1577884  230001710         101900_101902  0A421D7B11EABFC5        100.0   
4  1432367  230000063         102400_102401  B4734117F35EE97F        103.0   

      app_id  app_paid  campaign_id  carrier             city  \
0  2089229.0     False      1000023        1  137103102105100   
1  2070079.0     False      1000023        3  137105101100100   
2  2089397.0     False      1000021        3  137103104111100   
3  2071234.0     False      1003544        0  137103102113100   
4  1030051.0     False      1000023        1  137103102109100   

                         ...                            make            model  \
0                        ...                          HUAWEI  HUAWEI-CAZ-AL10   
1                        ...                          Xiaomi     Redmi Note 4   
2                        ...                            OPPO        OPPO+R11s   
3                        ...                             NaN         OPPO A57   
4                        ...                           Apple         iPhone 7   

   nnt  orderid  os  os_name     osv         province        time  \
0    1  3010798   2  android   7.0.0  137103102100100  2190219034   
1    1  2311397   2  android     6.0  137105101100100  2190221070   
2    1  3008491   2  android   7.1.1  137103104100100  2190219793   
3    1  3011304   2  android   6.0.1  137103102100100  2190221704   
4    3  2304491   1      ios  11.1.1  137103102100100  2190220024   

                                           user_tags  
0                                                NaN  
1  2100191,2100078,3001825,,3001781,3001791,30017...  
2                                                NaN  
3  2100098,gd_2100000,3001791,3001795,3002193,300...  
4                                                NaN  

[5 rows x 35 columns]

In [4]:
train.shape, test.shape, data.shape

((1001650, 35), (40024, 34), (1041674, 35))

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1041674 entries, 0 to 1041673
Data columns (total 35 columns):
adid                     1041674 non-null int64
advert_id                1041674 non-null int64
advert_industry_inner    1041674 non-null object
advert_name              1041674 non-null object
app_cate_id              1039376 non-null float64
app_id                   1039376 non-null float64
app_paid                 1041674 non-null bool
campaign_id              1041674 non-null int64
carrier                  1041674 non-null int64
city                     1041674 non-null int64
click                    1001650 non-null float64
creative_has_deeplink    1041674 non-null bool
creative_height          1041674 non-null int64
creative_id              1041674 non-null int64
creative_is_download     1041674 non-null bool
creative_is_js           1041674 non-null bool
creative_is_jump         1041674 non-null bool
creative_is_voicead      1041674 non-null bool
creative_tp_dnf      

In [7]:
data['f_channel'].count()

79777

# 3、初探
纠结，大佬说缺失值也是一种特征，所以缺失值使用-1填充，最后训练的时候使用 原始特征+新添的特征
## 缺失值

In [26]:
cnt = data.shape[0]   # 总的样本数量
print('训练样本', train.shape[0])
print('测试样本', test.shape[0])
print('总样本', cnt)

for feat in data.columns:
    tmp = cnt - data[feat].count()
    if tmp > 0:
        print(feat, tmp, '\t%.3f'%(tmp / cnt * 100), '%')

训练样本 1001650
测试样本 40024
总样本 1041674
app_cate_id 2298 	0.221 %
app_id 2298 	0.221 %
click 40024 	3.842 %
f_channel 961897 	92.341 %
make 103043 	9.892 %
model 7793 	0.748 %
osv 8211 	0.788 %
user_tags 323002 	31.008 %


- app_cate_id app分类 2298  
- app_id 媒体id 2298  
- f_channel 一级频道 961897  （缺失太多）  
- make 品牌 103043  
- model 机型 7793 
- osv 操作系统 8211
- user_tags 用户标签信息 323002

In [42]:
pd.unique(data.dtypes)

array([dtype('int64'), dtype('O'), dtype('float64'), dtype('bool')],
      dtype=object)

## 去重

In [84]:
data.drop_duplicates(subset=None, keep='first', inplace=False)

adid  advert_id       advert_name  app_cate_id     app_id  \
0        1560128  230000063  B4734117F35EE97F        107.0  2089229.0   
1        1488859  230000063  B4734117F35EE97F        108.0  2070079.0   
2        1537089  230000065  E257895F74792E81        100.0  2089397.0   
3        1577884  230001710  0A421D7B11EABFC5        100.0  2071234.0   
4        1432367  230000063  B4734117F35EE97F        103.0  1030051.0   
5        1559158  230000833  862FF2E9B0AD4C14        103.0  2071267.0   
6        1580496  230001696  310405E93895BD58        100.0  2071133.0   
7        1560128  230000063  B4734117F35EE97F        107.0  2089229.0   
8        1577884  230001710  0A421D7B11EABFC5        100.0  2071234.0   
9        1579122  230000063  B4734117F35EE97F        100.0  2089397.0   
10       1579878  230001696  310405E93895BD58        100.0  2071133.0   
11       1580588  230000245  68C5520FFBC5FC19        111.0  2070889.0   
12       1548446  230001025  50822F8310CAE01D        108.0  2070463.0   
13       1577884  230001710  0A421D7B11EABFC5        100.0  2071234.0   
14       1573112  230000065  E257895F74792E81        103.0  2071103.0   
15       1580394  230000245  68C5520FFBC5FC19        103.0  1030051.0   
16       1580580  230000245  68C5520FFBC5FC19        108.0  1080008.0   
17       1574082  230000063  B4734117F35EE97F        107.0  2089303.0   
18       1548194  230000063  B4734117F35EE97F        100.0  2089397.0   
19       1579858  230001696  310405E93895BD58        100.0  2071133.0   
20       1580210  230001696  310405E93895BD58        100.0  2071133.0   
21       1451143  230000063  B4734117F35EE97F        105.0  2070241.0   
22       1580210  230001696  310405E93895BD58        100.0  2071133.0   
23       1575260  230001694  5C84C3D2EBBB9D7C        111.0  2070351.0   
24       1577884  230001710  0A421D7B11EABFC5        100.0  2071234.0   
25       1579834  230001696  310405E93895BD58        111.0  2070351.0   
26       1580370  230001696  310405E93895BD58        100.0  2071133.0   
27       1575104  230001547  37E91828E6FB695A        108.0  2090600.0   
28       1547080  230000065  E257895F74792E81        107.0  2070969.0   
29       1575100  230000063  B4734117F35EE97F        111.0  1110016.0   
...          ...        ...               ...          ...        ...   
1041644  1586162  230001025  50822F8310CAE01D        107.0  2070969.0   
1041645  1577888  230001710  0A421D7B11EABFC5        100.0  2071234.0   
1041646  1456557  230000063  B4734117F35EE97F        107.0  1100001.0   
1041647  1517657  230000063  B4734117F35EE97F        108.0  1080008.0   
1041648  1456557  230000063  B4734117F35EE97F        107.0  1100001.0   
1041649  1547080  230000065  E257895F74792E81        107.0  2070969.0   
1041650  1549532  230000063  B4734117F35EE97F        100.0  2089397.0   
1041651  1585512  230000063  B4734117F35EE97F        107.0  2089229.0   
1041652  1585884  230000063  B4734117F35EE97F        107.0  2089229.0   
1041653  1549584  230000065  E257895F74792E81        100.0  2089397.0   
1041654  1585966  230001025  50822F8310CAE01D        110.0  2070523.0   
1041655  1585512  230000063  B4734117F35EE97F        107.0  2089229.0   
1041656  1495325  230000063  B4734117F35EE97F        108.0  2071279.0   
1041657  1579370  230001620  CC4602621BE800EA        100.0  2089397.0   
1041658  1547080  230000065  E257895F74792E81        107.0  2070969.0   
1041659  1585436  230000063  B4734117F35EE97F        100.0  2089397.0   
1041660  1584418  230001025  50822F8310CAE01D        108.0  2071003.0   
1041661  1577888  230001710  0A421D7B11EABFC5        100.0  2071234.0   
1041662  1585724  230000063  B4734117F35EE97F        105.0  2090638.0   
1041663  1585512  230000063  B4734117F35EE97F        107.0  2089229.0   
1041664  1585884  230000063  B4734117F35EE97F        107.0  2089229.0   
1041665  1575398  230001636  42A4CB9035B7F50E        111.0  1110016.0   
1041666  1577888  230001710  0A421D7B11EABFC5        100.0  2071234.0 

## 删除f_channel 一级频道特征

In [45]:
data.drop('f_channel', axis=1, inplace=True)

## 提取字符类型特征

In [47]:
from pandas.api.types import is_string_dtype

for feat in data.columns.tolist():
    if is_string_dtype(data[feat]):
        print(feat)

advert_industry_inner
advert_name
inner_slot_id
make
model
os_name
osv
user_tags


## 特征工程
补充说明：  
1. advert_industry_inner字段数据样例为102400_102401，“102400(前者)”表示广告主一级行业标签id，“102401(后者)”表示广告主二级行业id，如“教育_培训”
2. time字段脱敏后为有序排列，且时间间隔和与真实时间对应。

1、advert_industry_inner拆分开  
2、advert_name都是一些B4734117F35EE97F字符串，可以考虑`删除`或者进行pd.get_dummies处理  
3、inner_slot_id媒体广告位也是字符串xf_275C061483984E075832A4373BDDF27B，可以考虑取‘_’前的字段  
4、make model品牌机型可以考虑给机型加上品牌前缀， os_name osv操作系统和版本同理  
5、user_tags用户标签没有什么可以提取的

In [49]:
data['advert_industry_inner_0'] = data['advert_industry_inner'].apply(lambda x: x.split('_')[0])
data['advert_industry_inner_1'] = data['advert_industry_inner'].apply(lambda x: x.split('_')[1])
data['inner_slot_id_0'] = data['inner_slot_id'].apply(lambda x:x.split('_')[0])
data['model'] = data['make'].astype(str).values + '_' + data['model'].astype(str).values
data['osv'] = data['os'].astype(str).values + '_' + data['osv'].astype(str).values

data.drop(['advert_industry_inner', 'advert_name', 'inner_slot_id', 'make', 'model', 'os', 'osv'], axis=1, inplace=True)

In [50]:
data.shape

(1041674, 33)

## 提取bool布尔类型特征
转换为int类型

In [83]:
from pandas.api.types import is_bool_dtype

for feat in data.columns.tolist():
    if is_bool_dtype(data[feat]):
        print(feat)
        data[feat] = data[feat].astype(int)

app_paid
creative_has_deeplink
creative_is_download
creative_is_js
creative_is_jump
creative_is_voicead


## 时间time
补充说明：  
1. advert_industry_inner字段数据样例为102400_102401，“102400(前者)”表示广告主一级行业标签id，“102401(后者)”表示广告主二级行业id，如“教育_培训”
2. time字段脱敏后为有序排列，且时间间隔和与真实时间对应。

数据集提供了若干天的样本，最后一天数据构成了测试集，其余日期的数据作为训练数据。

所以年份就不需要了。

In [40]:
import time
day = data['time'].apply(lambda x: int(time.strftime("%d", time.localtime(x))))
hour = data['time'].apply(lambda x: int(time.strftime("%H", time.localtime(x))))
day.head()

0    29
1    29
2    29
3    29
4    29
Name: time, dtype: int64

In [39]:
time.localtime(2190219034)

time.struct_time(tm_year=2039, tm_mon=5, tm_mday=29, tm_hour=2, tm_min=10, tm_sec=34, tm_wday=6, tm_yday=149, tm_isdst=0)

## city特征

In [77]:
head = []
tail = []
for city in data['city']:
    h = str(city)[0:5]
    t = str(city)[-3:]
    head.append(h)
    tail.append(t)
print(set(head), set(tail))

{'0', '13710'} {'100', '0'}


结论：
1. 前5个数字都是一样的，尾部3个数字都是一样的，无效。
2. city中数字为0。
3. 总的长度是15。
4. 城市：国家、省份、城市、等级
5. 基于4，剩下7个数字，等级占1个，其余占2个，即5-7为国家、5-9为省份、5-11为城市、11-12为等级

In [74]:
data['city'][117]

0

In [78]:
data['city'][0:5]

0    137103102105100
1    137105101100100
2    137103104111100
3    137103102113100
4    137103102109100
Name: city, dtype: int64

In [79]:
data['county'] = data['city'].apply(lambda x: str(x)[5:7])
data['province'] = data['city'].apply(lambda x: str(x)[5:9])
data['city'] = data['city'].apply(lambda x: str(x)[5:11])
data['level'] = data['city'].apply(lambda x: str(x)[11:12])

In [48]:
pd.set_option('display.max_columns', None)    # 强制显示所有属性值
data.head()

adid  advert_id advert_industry_inner       advert_name  app_cate_id  \
0  1560128  230000063         102400_102401  B4734117F35EE97F        107.0   
1  1488859  230000063         102400_102401  B4734117F35EE97F        108.0   
2  1537089  230000065         101700_101704  E257895F74792E81        100.0   
3  1577884  230001710         101900_101902  0A421D7B11EABFC5        100.0   
4  1432367  230000063         102400_102401  B4734117F35EE97F        103.0   

      app_id  app_paid  campaign_id  carrier             city  click  \
0  2089229.0     False      1000023        1  137103102105100    0.0   
1  2070079.0     False      1000023        3  137105101100100    0.0   
2  2089397.0     False      1000021        3  137103104111100    0.0   
3  2071234.0     False      1003544        0  137103102113100    0.0   
4  1030051.0     False      1000023        1  137103102109100    0.0   

   creative_has_deeplink  creative_height  creative_id  creative_is_download  \
0                  False              720      2338420                 False   
1                  False              640      2310417                 False   
2                  False              640      2337017                 False   
3                  False              720      2342152                 False   
4                  False              640      2305409                 False   

   creative_is_js  creative_is_jump  creative_is_voicead      creative_tp_dnf  \
0           False              True                False  8390267734059046014   
1           False              True                False  8390208550469153745   
2           False              True                False  8390430283595430291   
3           False              True                False  8390229093704413749   
4           False              True                False  8390208550469153745   

   creative_type  creative_width  devtype  \
0              8            1280        2   
1              8             960        2   
2              8             960        2   
3              3            1280        2   
4              8             960        2   

                         inner_slot_id          instance_id    make  \
0  xf_275C061483984E075832A4373BDDF27B    86294719979897807  HUAWEI   
1  xf_D84DAB691E2E08C5B80D2FF5135F886E  2699289844928136052  Xiaomi   
2  xf_7F9FF3BEA11FE5B3AE6332EFBBD59496  3117527168445845752    OPPO   
3            iqy_1000000000381-1-15-15  3398484891050993371     NaN   
4  xf_6C4DCB36DBE7EB12CE55EDF319FF8D93  2035477570591176488   Apple   

             model  nnt  orderid  os  os_name     osv         province  \
0  HUAWEI-CAZ-AL10    1  3010798   2  android   7.0.0  137103102100100   
1     Redmi Note 4    1  2311397   2  android     6.0  137105101100100   
2        OPPO+R11s    1  3008491   2  android   7.1.1  137103104100100   
3         OPPO A57    1  3011304   2  android   6.0.1  137103102100100   
4         iPhone 7    3  2304491   1      ios  11.1.1  137103102100100   

         time                                          user_tags  
0  2190219034                                                NaN  
1  2190221070  2100191,2100078,3001825,,3001781,3001791,30017...  
2  2190219793                                                NaN  
3  2190221704  2100098,gd_2100000,3001791,3001795,3002193,300...  
4  2190220024                                                NaN

In [127]:
# 新添的特征
addFeature = ['advert_industry_inner_0', 'inner_slot_id_0',  'county', 'level']
# 特征选择
advertFeature = ['adid', 'advert_id', 'orderid',  'advert_industry_inner', 'advert_name', 'campaign_id', 'creative_id', 'creative_type', 'creative_tp_dnf', 'creative_has_deeplink', 'creative_is_jump', 'creative_is_download']  # 没有后4个特征
mediaFeature = ['app_cate_id', 'f_channel', 'app_id', 'inner_slot_id']   # 没有app_paid
contentFeature = ['city', 'carrier', 'province', 'nnt', 'devtype', 'os_name', 'osv', 'os', 'make', 'model']  # 没有time
otherFeature = ['creative_width', 'creative_height', 'hour']

feature = advertFeature + mediaFeature + contentFeature + addFeature
len(feature)

30

In [ ]:
减少 time\user_tags\app_paid\creative_is_js\creative_is_voicead\creative_width\creative_height\instance_id   8
增加 adid_prefix\level\couty\inner_slot_prefix\advert_industry_inner_1   5